### PR TITLE
 ENT-2880, ENT 2881 | Updated CouponRedeemView and VoucherAddLogicMix…

### PR DESCRIPTION
**Pull request by HammadAhmadWaqas**
…in to support multiple products to redeem using vouchers including programs.

After this PR we should be able to use coupons on the program enrollments.
(edx-enterprise version 3.3.4 or later is required)

### CODE REVIEW

File: ecommerce/courses/utils.py [hunk 1]
```diff
@@ -23,6 +30,44 @@ def mode_for_product(product):
...
+                )
+                return None
+            else:
+                course_ids.append(course_key)
+        elif product.is_course_entitlement_product:
+            try:
+                response = get_course_info_from_catalog(site, product)
+            except (ReqConnectionError, AttributeError, SlumberHttpBaseException, Timeout) as exc:
+                logger.exception(
+                    'Unable fetch course ids because products contains a course entitlement product but we failed to'
```
Comment (hasnain-naveed) [inline ecommerce/courses/utils.py]: This logger message has some issues. Can we please simplify this?

File: ecommerce/coupons/views.py [hunk 2]
```diff
@@ -205,58 +236,71 @@ def get(self, request):  # pylint: disable=too-many-statements
...
-                               'course in the program. If you need assistance, contact edX support.')
-                }
-            )
+        if enterprise_customer:
+            sku_to_course_id_map = self._get_sku_to_course_id_map(skus, products, request.site)
+            if not sku_to_course_id_map:
+                return render(
+                    request,
+                    template_name,
+                    {'error': _('Could not find course_ids for few of the products.')}
```
Comment (hasnain-naveed) [inline ecommerce/coupons/views.py]: can we add some elaborative message ?

File: ecommerce/coupons/views.py [hunk 3]
```diff
@@ -205,58 +236,71 @@ def get(self, request):  # pylint: disable=too-many-statements
...
+                        failure_url=failure_url
+                    )
+                    return HttpResponseRedirect(redirect_url)
 
         try:
-            basket = prepare_basket(request, [product], voucher)
+            basket = prepare_basket(request, products, voucher)
         except AlreadyPlacedOrderException:
-            msg = _('You have already purchased {course} seat.').format(course=product.course.name)
+            msg = _('You have already purchased these courses.')
```
Comment (hasnain-naveed) [inline ecommerce/coupons/views.py]: Can we add the product title here as it was added previously?

File: ecommerce/coupons/views.py [hunk 4]
```diff
@@ -146,6 +147,33 @@ def get(self, request, *args, **kwargs):
...
+        return False
+
+    def _get_sku_to_course_id_map(self, skus, products, site):
+        """Returns tuples containing sku and relevant course_id"""
+        sku_to_course_id_map = []
+        for sku in skus:
+            product = products.filter(stockrecords__partner_sku=sku).first()
+            try:
+                course_id = get_products_course_ids(site, [product])[0]
+            except (TypeError, AttributeError):
```
Comment (hasnain-naveed) [inline ecommerce/coupons/views.py:170]: Curious why are we catching `AttributeError` exception here when you are catching this in child function `get_products_course_ids `?

File: ecommerce/coupons/views.py [hunk 5]
```diff
@@ -170,16 +199,18 @@ def get(self, request):  # pylint: disable=too-many-statements
...
-        try:
-            product = StockRecord.objects.get(partner_sku=sku).product
-        except StockRecord.DoesNotExist:
-            return render(request, template_name, {'error': _('The product does not exist.')})
+        products = Product.objects.filter(
+            stockrecords__partner_sku__in=skus,
+            stockrecords__partner=site_configuration.partner
+        )
+        if products.count() != len(skus):
+            return render(request, template_name, {'error': _('These products do not exist.')})
```
Comment (hasnain-naveed) [inline ecommerce/coupons/views.py]: This is a very ambiguous message for the end-user.

**Review by hasnain-naveed — CHANGES_REQUESTED** [review]

File: ecommerce/coupons/views.py (hunk 5 above)
Comment (HammadAhmadWaqas) [reply ecommerce/coupons/views.py]: `The courses in which you are trying to enroll, not found.`

File: ecommerce/coupons/views.py [hunk 6]
```diff
@@ -170,16 +199,18 @@ def get(self, request):  # pylint: disable=too-many-statements
...
 
-        valid_voucher, msg = voucher_is_valid(voucher, [product], request)
+        valid_voucher, msg = voucher_is_valid(voucher, products, request)
         if not valid_voucher:
-            logger.warning('[Code Redemption Failure] The voucher is not valid for this product. '
-                           'User: %s, Product: %s, Code: %s, Message: %s',
-                           request.user.username, product.id, voucher.code, msg)
+            logger.warning('[Code Redemption Failure] The voucher is not valid for these products. '
```
Comment (HammadAhmadWaqas) [inline ecommerce/coupons/views.py]: `[Code Redemption Failure] The voucher is not valid for products. `
will use pluralize function.

File: ecommerce/coupons/views.py (hunk 2 above)
Comment (HammadAhmadWaqas) [reply ecommerce/coupons/views.py]: `There is some internal error. please contact your admin.`

File: ecommerce/courses/utils.py (hunk 1 above)
Comment (HammadAhmadWaqas) [reply ecommerce/courses/utils.py]: `Supported Entitlement Product could not be fetched from the discovery catalog.`

File: ecommerce/coupons/views.py (hunk 4 above)
Comment (HammadAhmadWaqas) [reply ecommerce/coupons/views.py:170]: becasue `get_products_course_ids` can raise `AttributeError` if the `product` in None

File: ecommerce/coupons/views.py (hunk 3 above)
Comment (HammadAhmadWaqas) [reply ecommerce/coupons/views.py]: there was a single product before. now there could be multiple.

File: ecommerce/coupons/views.py [hunk 7]
```diff
@@ -146,6 +148,33 @@ def get(self, request, *args, **kwargs):
 class CouponRedeemView(EdxOrderPlacementMixin, APIView):
     permission_classes = (LoginRedirectIfUnauthenticated,)
 
+    def _is_valid_consent_token(self, received_consent_token, access_token, enterprise_customer_id, course_ids):
+        """Iterate over all the course_id and check if received_consent_token in valid for any of the course."""
```
Comment (saleem-latif) [inline ecommerce/coupons/views.py]: nit:
```Iterate over all the `course_id`s and check if `received_consent_token` is valid for any of the courses.```

File: ecommerce/coupons/views.py [hunk 8]
```diff
@@ -146,6 +148,33 @@ def get(self, request, *args, **kwargs):
...
+        for sku in skus:
+            product = products.filter(stockrecords__partner_sku=sku).first()
+            try:
+                course_id = get_products_course_ids(site, [product])[0]
+            except (TypeError, AttributeError):
+                logger.exception(
+                    "[Code Redemption Failure] could not find course id. Product: %s, sku: %s", product, sku
+                )
+                return None
+            sku_to_course_id_map.append((sku, course_id))
```
Comment (saleem-latif) [inline ecommerce/coupons/views.py]: Wouldn't it be better to use a dictionary here instead of a list of tuples?

File: ecommerce/courses/utils.py [hunk 9]
```diff
@@ -23,6 +30,43 @@ def mode_for_product(product):
...
+                    site,
+                    product,
+                    exc
+                )
+                return None
+            else:
+                course_ids.append(course_key)
+        elif product.is_course_entitlement_product:
+            try:
+                response = get_course_info_from_catalog(site, product)
```
Comment (saleem-latif) [inline ecommerce/courses/utils.py:53]: Is there a workaround for making an API call inside the loop?

**Review by saleem-latif — APPROVED** [review]
Few comments, other than that Looks Good :+1:

File: ecommerce/courses/utils.py (hunk 9 above)
Comment (HammadAhmadWaqas) [reply ecommerce/courses/utils.py:53]: for entitlement products, we have to get the course key from catalog. As it is a cached API, it should not make any trouble.

**Comment by hasnain-naveed** [comment]
@HammadAhmadWaqas we need an extra pair of eyes for this PR as this will change the core functionality of coupons. We should tag someone from the Titan team. From my side, its good to go. But we should add a waffle flag to enable/disable this functionality.  👍

**Review by hasnain-naveed — APPROVED** [review]

**Review by moconnell1453 — APPROVED** [review]

**Review by saleem-latif — APPROVED** [review]

**Pull request closed without merging**